### ISSUE TITLE
[Bug] DeepSeek V4 8-card serving 性能验证与瓶颈分析

### ISSUE THREAD

**Issue by high-cloud** [hw-native-sys/pypto-serving#95]
# DeepSeek V4 8-card Serving 性能验证与瓶颈分析

## 一、验证概述

本文记录 DeepSeek V4 Flash W8A8 在 8 卡 Ascend NPU 上的 serving 性能基线，并拆分 prefill、decode、runtime bind/H2D、Device 执行、D2H、LM head 和 sampling 等阶段耗时。

本报告使用当前尚未包含权重和 Cache 驻留优化的基线版本，不包含未合入 PR 的优化后性能数据。

验证范围：

- DeepSeek V4 Flash W8A8
- TP=8，DP=1
- MTP disabled
- Prompt：`Huawei is`，共 4 tokens
- 生成 20 tokens
- 完整 serving 推理
- Runtime STRACE
- 不使用泳道图作为耗时统计依据

相关 issue：

- #44：逐 token decode 输入 tensor 的 Device buffer 池化
- #62：DeepSeek V4 Flash serving 接入

本 issue 的重点是当前完整推理路径的性能分布，以及权重、Cache 和运行时资源生命周期优化。

## 二、测试环境

| 项目 | 配置 |
| --- | --- |
| Serving commit | `750a65918a2ffef8c15935b4b127637f27ca073c` |
| pypto-lib commit | `306519dcac1aa0d9edf273299cc0648c4835ec46` |
| Model | `/data/models/dsv4-flash-w8a8` |
| Quantization | W8A8 |
| Platform | `a2a3` |
| Devices | `8,9,10,11,12,13,14,15` |
| TP / DP | TP=8 / DP=1 |
| PTOAS | `0.48` |
| MTP | Disabled |
| Prompt | `Huawei is` |
| Prompt tokens | 4 |
| Generated tokens | 20 |
| Task | `task_20260715_194739_142047818271` |
| Result | Exit 0，生成文本符合预期 |

### 3.1 指标定义

本次统计包含以下层级：

1. **Serving wall**

   从 scheduler 发起一个模型 step，到 sampling 完成的 Host wall。

2. **Packed dispatch outer wall**

   L3 父进程提交 8 个 rank、等待所有 rank 返回的整体 wall time。

3. **Runtime Host wall**

   每个 rank 的 `simpler_run` wall time。

4. **Bind/H2D**

   Runtime 为 callable 参数分配 Device 内存并执行 H2D 的耗时，包括模型权重、Cache 和调用参数。

5. **Device wall**

   Device graph/orchestrator/scheduler 的临界路径 wall time。三个窗口高度重叠，不能相加。

6. **Validate/D2H**

   Runtime 检查输出、执行必要的 D2H 并回收本次调用资源的耗时。

7. **Host LM head / sampling**

   Packed model 返回 hidden state 后，在 Host 侧进行的 LM head 和 sampling。

### 3.2 Prefill 性能

| 阶段 | 耗时 |
| --- | ---: |
| Serving prefill total | 444297.694 ms |
| Batch planning | 0.022 ms |
| Tensorize inputs | 0.920 ms |
| Embedding lookup | 2.301 ms |
| Model runner | 444281.024 ms |
| Shared-buffer first allocation | 368028.127 ms |
| Prepare model inputs | 49.649 ms |
| Stage kernel inputs | 83.342 ms |
| Packed prefill dispatch outer wall | 75810.681 ms |
| Output checks | 0.165 ms |
| Host LM head | 307.303 ms |
| Sampling | 13.346 ms |

Prefill 的主要耗时不是 kernel 本身，而是首次申请和初始化共享 Host buffer：

- Shared-buffer first allocation：368.028 s，占 prefill 总耗时约 82.83%。
- Packed prefill dispatch：75.811 s，占约 17.06%。
- LM head、sampling、embedding 等阶段占比较小。

这部分属于明显的冷启动成本，建议将固定 shape buffer 的申请和初始化前移到模型加载阶段。

### 3.3 Steady Decode 性能

以下统计包含 18 个正常 decode token，不包含开启额外 profiling 的第一个 decode token。

| 阶段 | Mean | P50 | P90 | Min | Max |
| --- | ---: | ---: | ---: | ---: | ---: |
| Serving decode total | 7488.817 ms | 7445.914 ms | 8018.532 ms | 6596.584 ms | 9067.067 ms |
| Batch planning | 0.035 ms | 0.034 ms | 0.043 ms | 0.026 ms | 0.050 ms |
| Tensorize inputs | 0.126 ms | 0.118 ms | 0.171 ms | 0.073 ms | 0.290 ms |
| Embedding lookup | 1.337 ms | 0.326 ms | 3.837 ms | 0.202 ms | 4.391 ms |
| Model runner | 7461.152 ms | 7407.855 ms | 8001.793 ms | 6582.390 ms | 9051.831 ms |
| Prepare/stage inputs | 37.320 ms | 26.703 ms | 88.988 ms | 19.283 ms | 91.347 ms |
| Stage packed-kernel args | 6.068 ms | 4.208 ms | 19.991 ms | 3.285 ms | 21.136 ms |
| Packed decode dispatch outer wall | 7184.023 ms | 7106.150 ms | 7796.148 ms | 6288.459 ms | 8777.045 ms |
| Output checks | 0.105 ms | 0.105 ms | 0.121 ms | 0.085 ms | 0.137 ms |
| Host LM head | 233.045 ms | 195.019 ms | 386.988 ms | 147.294 ms | 559.913 ms |
| Sampling | 26.139 ms | 13.562 ms | 71.406 ms | 9.614 ms | 90.752 ms |

Packed decode dispatch 平均为 7184.023 ms，占 serving decode 总耗时约 95.93%，是当前最主要的 serving 瓶颈。

Packed dispatch 之后的 Host LM head 和 sampling 合计平均约 259.184 ms，占约 3.46%。

### 3.4 Runtime STRACE

共统计 19 个 decode token × 8 个 rank，即 152 次 rank runtime 调用。

| Runtime 阶段 | 每个 rank 调用平均耗时 | Host wall 占比 |
| --- | ---: | ---: |
| `simpler_run` Host wall | 6906.775 ms | 100% |
| Bind total | 5231.321 ms | 75.74% |
| Bind args：Device allocation/H2D | 5231.202 ms | 75.74% |
| Bind prebuilt | 0.013 ms | <0.01% |
| Device wall | 1365.794 ms | 19.78% |
| Output validation/D2H | 278.013 ms | 4.03% |
| 其他 Host 开销 | 约 31.647 ms | 0.46% |

Device 内部阶段：

| Device 阶段 | 平均耗时 |
| --- | ---: |
| Device graph window | 1365.762 ms |
| Device orchestrator window | 1365.659 ms |
| Device scheduler window | 1365.734 ms |
| Device preamble | 0.024 ms |
| Device config validation | 0.005 ms |
| Device arena wiring | 0.003 ms |
| Device SM reset | 0.004 ms |
| Device post-orchestration | 0.007 ms |

Device graph、orchestrator 和 scheduler 是同一执行区间的不同观察窗口，三者高度重叠，不能相加。当前 Device 有效临界路径约为 1.366 s。

每个 token 最慢 rank 的平均数据：

| 指标 | 耗时 |
| --- | ---: |
| Slowest-rank Host wall | 6978.185 ms |
| Slowest-rank bind | 6466.179 ms |
| 独立选择的最大 Device wall | 2534.020 ms |

最大 Host wall、bind 和 Device wall 可能来自不同 rank，因此这些最大值也不能直接相加。

#### 1. 权重和 Cache 每个 decode step 重复 bind/H2D

Bind/H2D 平均耗时 5231.202 ms，占单 rank runtime Host wall 的 75.74%。

当前每个 decode token 都重新处理固定模型参数和长生命周期 Cache。对于 decode 来说，模型权重不变，KV Cache、compressor Cache 等状态也应该由 Device 上的长期 tensor 原地更新，不应每个 step 重新申请和搬运。

在其他阶段耗时保持不变的理想情况下，如果 steady decode 消除这部分重复 bind/H2D，单 rank runtime wall 的理论下限可由 6906.775 ms 降到约 1675.573 ms。该数值只是根据基线拆分计算的优化空间，并非优化后实测结果。

#### 2. Device 执行仍有较大优化空间

消除 bind/H2D 后，Device wall 将成为主要瓶颈，目前平均为 1365.794 ms。

Device 侧应进一步检查：

- MoE dispatch/wait 中的跨 rank 同步；
- rank 间负载不均衡；
- 通信 domain 和通信资源是否逐 step 创建、销毁；
- scheduler/orchestrator 是否存在不必要的同步；
- main decode kernel 的实际有效执行时间与 Device wall 之间的差异。

#### 3. Packed outer wall 存在额外父进程开销

Packed decode dispatch outer wall 为 7184.023 ms，而每个 token 的 slowest-rank Host wall 平均为 6978.185 ms，两者相差约 205.838 ms。

因为 L3 outer wall 和子进程 rank timing 的起止点、并发提交方式不同，这个差值不能简单归到某一个 kernel，但说明父进程仍存在约 200 ms 量级的资源管理、IPC、rank 提交或等待开销。

#### 4. Host LM head 和 sampling

Host LM head 平均 233.045 ms，sampling 平均 26.139 ms，合计约 259.184 ms。

在 runtime bind/H2D 优化后，这部分会成为明显的 serving 瓶颈。建议评估：

- LM head Device 化；
- sampling Device 化；
- hidden、logits 和 sampled token 保持 Device resident；
- sampled token 直接进入下一轮 embedding，避免 Device → Host → Device 往返。

#### 5. Prefill 冷启动申请

Prefill 首次共享 buffer 申请耗时 368028.127 ms。

建议将固定 shape 的 Host shared buffer、Device 权重、Cache、通信 domain 和 runtime arena 在模型启动阶段一次性初始化。请求执行阶段只更新实际输入数据和动态 metadata。

### 3.6 优化建议与验证目标

| 优先级 | 优化项 | 当前对应耗时 | 建议 |
| --- | --- | ---: | --- |
| P0 | 主模型权重 Device resident | Bind/H2D 5231.202 ms/rank | 模型加载时上传一次，decode 不再重复 H2D |
| P0 | KV/compressor Cache Device resident | 包含在 Bind/H2D 中 | 请求生命周期内保持 Device tensor，kernel 原地更新 |
| P0 | MTP 权重和 Cache resident | MTP 场景会产生同类开销 | 与主模型使用一致的生命周期管理 |
| P1 | 通信 domain/buffer 复用 | Outer 与 slowest rank 相差约 205.838 ms | Worker/runtime 生命周期内复用，避免逐 step 创建和释放 |
| P1 | Device LM head/sampling | 259.184 ms/token | 保持 hidden/logits/token 全流程 Device resident |
| P1 | MoE 通信和 rank 均衡 | Device wall 1365.794 ms | 优化 dispatch/wait 和跨 rank 同步 |
| P2 | Prefill buffer 启动时预分配 | 368028.127 ms cold-start | 将固定 buffer 初始化移出请求路径 |
| P2 | Decode 输入 tensor 池化 | 与 #44 互补 | 复用 hidden、seq_lens、block_table、slot_mapping 的 Device buffer |

建议后续在相同设备、相同 prompt 和相同 20-token 场景下验证：

- Steady decode 的固定权重和 Cache bind/H2D 接近 0；
- 不以增加每 token D2H 或 Host copy 为代价；
- Cache 原地更新与原始流程输出一致；
- 分别报告 main decode、MTP decode、runtime 和 serving wall；
- Packed outer wall 应逐步接近最慢 rank 的真实 Device 临界路径；
- 优化后数据单独记录，不覆盖本文的合入前基线。

**Comment by high-cloud** [comment]
## 2026-07-23 更新：Persistent L3 + Device LM head（无 MTP / MTP）

在最新 `origin/main` 基础上，使用 PR #111 的 persistent L3 与 profiling 改动重新测试。原 issue 正文中的优化前基线保留不变，本评论记录当前优化后结果。

### 测试环境

| 项目 | 配置 |
| --- | --- |
| Serving commit | `0a6cb04825fa35004b6ea0c04845c5ac97b11ba6`（PR #111） |
| Base commit | `c21760311567ee651c8aaf00ecdef5512dd8a66f` |
| pypto-lib commit | `c364bbcae44b4c17452576f4545081f1559c20fe` |
| Model | `/data/models/dsv4-flash-w8a8` |
| Platform / Devices | `a2a3` / `8,9,10,11,12,13,14,15` |
| DP / EP | 8 / 8 |
| PTOAS | `0.48` |
| Prompt | `Huawei is`（4 tokens） |
| Generated tokens | 20 |
| Profiling | `SA_PROFILE_LEVEL=verbose` |
| 无 MTP task | `task_20260723_063405_11649156745`，exit 0 |
| MTP task | `task_20260723_063405_116491432551`，exit 0 |

两组输出文本一致：

```text
 a leading global information and communications technology (ICT) solutions provider. Through our dedication to customer-centric innovation
```

### 端到端结果

| 指标 | 无 MTP | MTP |
| --- | ---: | ---: |
| 冷启动到 server healthy | 258.264 s | 540.566 s |
| HTTP completion wall | 7590.847 ms | 7539.834 ms |
| 20-token 请求吞吐 | 2.635 tok/s | 2.653 tok/s |
| Prefill worker wall | 1460.442 ms | 1434.283 ms |
| Decode step 数 | 19 | 10 |
| MTP acceptance | N/A | 9/10 = 90% |

20-token 短请求上，MTP 的完整 HTTP wall 仅降低约 **0.67%**。主要原因是首个 MTP decode step 需要初始化 draft；该步为 1143.228 ms，而后续 steady step 明显更快。

### Steady decode（剔除首个 decode step）

| 指标 | 无 MTP（18 steps） | MTP（9 steps） |
| --- | ---: | ---: |
| Worker decode step mean | 317.364 ms | 547.622 ms |
| Worker decode step P50 | 315.909 ms | 549.123 ms |
| Worker decode step P90 | 334.825 ms | 564.326 ms |
| 每 step 产出 token | 1 | 2 |
| 折算每输出 token | 317.364 ms | 273.811 ms |
| 折算 steady throughput | 3.151 tok/s | 3.652 tok/s |

MTP steady 阶段折算到每个输出 token 后，延迟降低约 **13.7%**，吞吐提高约 **15.9%**。生成长度增加后，首步 draft 初始化成本会被摊薄，MTP 收益会比本次 20-token 请求更明显。

### Runner / L3 span 拆分（steady mean）

| Span | 无 MTP | MTP |
| --- | ---: | ---: |
| `PyptoExecutor.run_decode` | 298.053 ms | 547.208 ms |
| Main `decode.l3_dispatch` | 233.864 ms | 314.245 ms |
| `decode.prepare` | 0.312 ms | 0.410 ms |
| `decode.prepare_inputs` | 13.631 ms | 8.650 ms |
| `decode.prepare_fwd_args` | 4.151 ms | 2.360 ms |
| `decode.mtp_advance` | N/A | 180.696 ms |
| MTP `decode.l3_dispatch` | N/A | 33.093 ms |
| MTP host LM head | N/A | 127.881 ms |

`decode.mtp_advance` 包含 MTP L3 dispatch、MTP host LM head 和约 19.7 ms 的其余状态处理，因此其子项不能与父 span 重复相加。

Main L3 在 MTP 下每 step 处理 2 个 token，step wall 从 233.864 ms 增加到 314.245 ms；折算后约为 157.123 ms/output-token。

### Serving 框架开销

| 开销 | 无 MTP | MTP |
| --- | ---: | ---: |
| Worker batch wrapper / sampling：`batch_decode - run_decode` | 19.311 ms/step | 0.414 ms/step |
| Queue/IPC wait：`scheduler.wait_worker_output - batch_decode` | 0.850 ms/step | 0.973 ms/step |
| `schedule + queue_worker_step + process_step_output` | 0.303 ms/step | 0.417 ms/step |
| HTTP wall 中未被 prefill/decode worker span覆盖的总开销 | 42.601 ms/request | 33.722 ms/request |

无 MTP 的约 19.3 ms worker wrapper 开销主要位于 executor 返回后 token sampling/result 处理；MTP 直接返回 accepted token IDs，因此该层开销明显更小。当前 scheduler、IPC 和 HTTP 框架开销已不是主要瓶颈。

### 与原始基线对比

原 issue 中无 MTP steady decode serving mean 为 7488.817 ms/token，packed decode dispatch 为 7184.023 ms/token。当前无 MTP steady worker decode 为 317.364 ms/token，main L3 dispatch 为 233.864 ms/token：

- steady decode wall 降低约 **95.8%**；
- main packed/L3 dispatch wall 降低约 **96.7%**。

这说明固定权重/Cache 重复 bind/H2D 与逐 step L3 生命周期成本已经基本从 steady hot path 移除。当前后续优化重点转为 main kernel、MTP host LM head，以及无 MTP 的 worker sampling/result 处理。
